### PR TITLE
Initial Classes

### PR DESCRIPTION
**Pull request by mecha**
The initial abstract classes and traits, together with default generic implementations: `Menu` and `MenuItem`.

### CODE REVIEW

**Comment by mecha** [comment]
Unit tests for `Menu` and `MenuItem` are currently incomplete.

**Comment by mecha** [comment]
A note on the hierarchy, which can be confusing since it attempts to keep both menus and menu items using a common ancestor.

All abstract functionality can be found in `MenuElementTrait`, `MenuTrait` and `MenuItemTrait`. There are also three abstract classes, each using one of these three traits, but only the `AbstractMenuElement` class is used in this package, which `AbstractBaseMenuElement` extends.

`AbstractBaseMenu` and `AbstractBaseMenuItem` then extend that, which are used to create the concrete implementations.

----

This is no longer true since recent refactoring. Now there is no distinction between a menu and a menu item. All menu nodes are `MenuElementInterface` instances, `ChildrenAwareNodeInterface` and `KeyAwareNodeInterface`.

File: src/AbstractMenuElement.php [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    protected function _hasParent()
+    {
+        return $this->_getParent() !== null;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
```
Comment (XedinUnknown) [inline src/AbstractMenuElement.php]: Looks like `@throws` is missing

File: src/AbstractMenuElement.php [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    /**
+     * Adds a child menu at a given position.
+     *
+     * @since [*next-version*]
+     *
+     * @param mixed    $child    The menu element to add as a child.
+     * @param int|null $position An integer, with larger values signifying a lower position, or null to append.
+     *
+     * @return $this
+     */
```
Comment (XedinUnknown) [inline src/AbstractMenuElement.php]: Looks like `@throws` is missing

File: src/AbstractMenuElement.php [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        if (!$this->_isMenuElement($child)) {
+            throw $this->_createValidationFailedException('Child is not a valid menu element.', 0, null, $child);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
```
Comment (XedinUnknown) [inline src/AbstractMenuElement.php]: Looks like `@throws` is missing.

File: src/AbstractMenuElement.php [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    /**
+     * Creates an exception for validation failure.
+     *
+     * @since [*next-version*]
+     *
+     * @param string                                      $message          The exception message.
+     * @param int                                         $code             The exception code.
+     * @param null                                        $inner            The previous exception in the chain.
+     * @param mixed                                       $subject          The subject that failed validation.
+     * @param string[]|StringableInterface[]|\Traversable $validationErrors The validation errors.
```
Comment (XedinUnknown) [inline src/AbstractMenuElement.php]: Could we make this a bit shorter?

File: src/AbstractMenuElement.php [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    protected function _validateParent($parent)
+    {
+        if ($parent !== null && !$this->_isMenuElement($parent)) {
+            throw $this->_createValidationFailedException('Parent is not a valid menu element.', 0, null, $parent);
```
Comment (XedinUnknown) [inline src/AbstractMenuElement.php]: Looks like no reasons for failed validation are given.

File: src/AbstractMenuElement.php [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    protected function _validateChild($child)
+    {
+        if (!$this->_isMenuElement($child)) {
+            throw $this->_createValidationFailedException('Child is not a valid menu element.', 0, null, $child);
```
Comment (XedinUnknown) [inline src/AbstractMenuElement.php]: Looks like no reasons for failed validation are given.

File: src/AbstractMenuElement.php [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+     * @return bool True if the given argument is a menu element instance, false if not.
+     */
+    protected function _isMenuElement($menuElement)
+    {
+        return $menuElement instanceof MenuElementInterface;
+    }
+
+    /**
+     * Normalizes the position.
+     *
```
Comment (XedinUnknown) [inline src/AbstractMenuElement.php]: A description of what happens would be great.

File: src/AbstractBaseMenuElement.php [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function getChildren()
+    {
+        return new RecursiveArrayIterator($this->_getChildren());
```
Comment (XedinUnknown) [inline src/AbstractBaseMenuElement.php]: ~Make factory method for this.~ Let's talk about recursive iteration first.

File: src/AbstractBaseMenuElement.php [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function hasChildren()
+    {
+        // What to do here? ChildrenAware::hasChildren() or RecursiveIterator::hasChildren() ?
```
Comment (XedinUnknown) [inline src/AbstractBaseMenuElement.php]: Yes, we have a problem. I also faced this problem with my `ConfigInterface` implementation. I have an idea of how to solve this using proper SoC. Let's talk about that ASAP.

**Review by XedinUnknown — CHANGES_REQUESTED** [review]
Documentation needs to be added about what kind of exceptions methods throw.
Also, the recursive iteration seems to be a problem to discuss in-depth.

File: src/AbstractMenuElement.php (hunk 1 above)
Comment (mecha) [reply src/AbstractMenuElement.php]: It's inherited from `ChildrenAwareTrait#_validateChild()`.

File: src/AbstractMenuElement.php (hunk 5 above)
Comment (mecha) [reply src/AbstractMenuElement.php]: Actually, menus are not meant to be parent-aware. I had forgotten to remove the trait.

File: src/AbstractMenuElement.php (hunk 4 above)
Comment (mecha) [reply src/AbstractMenuElement.php]: Don't know. Can we?

File: src/AbstractMenuElement.php (hunk 3 above)
Comment (mecha) [reply src/AbstractMenuElement.php]: Method has been removed. Reason: menu elements should not be parent-aware.

File: src/AbstractMenuElement.php (hunk 4 above)
Comment (XedinUnknown) [reply src/AbstractMenuElement.php]: E.g. by aliasing `StringableInterface` to `Stringable`, like I always do, and by `use`ing `Traversable`.

**Comment by XedinUnknown** [comment]
OMG the `class RecursiveMenuIterator extends AbstractBaseRecursiveIterator` is so amaze!!! @mecha, I cri of joi.

**Pull request merged**